File: java/std_nat/src/fan/std/FileBufPeer.java
package fan.std;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.RandomAccessFile;

import fan.sys.ByteArray;
import fan.sys.IOErr;

public class FileBufPeer {
	private RandomAccessFile fp;

	public FileBufPeer make(FileBuf buf) {
		return new FileBufPeer();
	}

	void init(FileBuf self, File file, String mode) throws FileNotFoundException {
		LocalFile lfile = (LocalFile) file;
		fp = new RandomAccessFile((java.io.File) lfile.peer, mode);
	}

	long size(FileBuf self) {
		try {
			return fp.length();
		} catch (IOException e) {
			throw IOErr.make(e);
		}
	}

	public final void size(FileBuf self, long x) {
		try {
			fp.setLength(x);
		} catch (IOException e) {
			throw IOErr.make(e);
		}
	}

	long capacity(FileBuf self) {
		return Long.MAX_VALUE;
	}
	
	void capacity(FileBuf self, long capa) {
	}

	long pos(FileBuf self) {
		try {
			return fp.getFilePointer();
		} catch (IOException e) {
			throw IOErr.make(e);
		}
	}

	void pos(FileBuf self, long x) {
		try {
			fp.seek(x);
		} catch (IOException e) {
			throw IOErr.make(e);
		}
	}

	long getByte(FileBuf self, long pos) {
		try {
			long oldPos = fp.getFilePointer();
			fp.seek(pos);
			int b = fp.read();
			fp.seek(oldPos);
			return b;
		} catch (IOException e) {
			throw IOErr.make(e);
		}
	}

	void setByte(FileBuf self, long pos, long b) {
		try {
			long oldPos = fp.getFilePointer();
			fp.seek(pos);
			fp.write((int) b);
			fp.seek(oldPos);
		} catch (IOException e) {
			throw IOErr.make(e);
		}
	}

	long getBytes(FileBuf self, long pos, ByteArray dst, long off, long len) {
		try {
			long oldPos = fp.getFilePointer();
			fp.seek(pos);
			int r = fp.read(dst.array(), (int)off, (int)len);
			fp.seek(oldPos);
			return r;
		} catch (IOException e) {
			throw IOErr.make(e);
		}
	}
	
	void setBytes(FileBuf self, long pos, ByteArray src, long off, long len) {
		try {
			long oldPos = fp.getFilePointer();
			fp.seek(pos);
			fp.write(src.array(), (int)off, (int)len);
			fp.seek(oldPos);
		} catch (IOException e) {
			throw IOErr.make(e);
		}
	}

	boolean close(FileBuf self) {
		try {
			fp.close();
			return true;
		} catch (Exception e) {
			return false;
		}
	}

	FileBuf sync(FileBuf self) {
		try {
			fp.getFD().sync();
			return self;
		} catch (IOException e) {
			throw IOErr.make(e);
		}
	}
}

File: src/std/js/FileBufPeer.js


fan.std.FileBufPeer.prototype.init = function(self, file, mode) {
	var reader = new FileReader();
	var buf = reader.readAsArrayBuffer(file);
	this.m_fp = new DataView(buf);
	this.m_pos = 0;
}

fan.std.FileBufPeer.prototype.size() = function(self) {
	this.fp.byteLength;
}

fan.std.FileBufPeer.prototype.size$() = function(self, x) {
}

fan.std.FileBufPeer.prototype.capacity() = function(self) {
	throw MAX_VALUE;
}

fan.std.FileBufPeer.prototype.capacity$() = function(self, capa) {
}

fan.std.FileBufPeer.prototype.pos() = function(self) {
	return this.m_pos;
}

fan.std.FileBufPeer.prototype.pos$() = function(self, x) {
	this.m_pos = x;
}

fan.std.FileBufPeer.prototype.getByte() = function(self, pos) {
	return this.m_pf.getUint8(pos)
}

fan.std.FileBufPeer.prototype.setByte() = function(self, pos, b) {
	this.m_pf.setUint8(pos, b);
}

fan.std.FileBufPeer.prototype.getBytes() = function(self, pos, dst, off, len) {
	var size = this.m_pf.byteLength-pos;
	if (size > len) size = len;

	for (var i=0; i<size; ++i) {
		dst.set(off+i, this.m_pf.getUint8(pos+i));
	}
	return size;
}

fan.std.FileBufPeer.prototype.setBytes() = function(self, pos, src, off, len) {
	var size = this.m_pf.byteLength-pos;
	if (size > len) size = len;

	for (var i=0; i<size; ++i) {
		this.m_pf.setUint8(pos+i, dst.get(off+i));
	}
	return size;
}

fan.std.FileBufPeer.prototype.close() = function(self) {
}

fan.std.FileBufPeer.prototype.sync() = function(self) {
}

